# Assignment 2 Rough Work

In [58]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC, SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, log_loss
from pandas.api.types import CategoricalDtype
%matplotlib inline

## 1. Exploratory Analysis

In [63]:
# create data frames
data_path = os.path.join(os.getcwd(), 'Data', 'PartA', '20ng_train.csv')
trn_newsgroup = pd.read_csv(data_path, delimiter=',', compression='bz2')
data_path2 = os.path.join(os.getcwd(), 'Data', 'PartA', '20ng_test.csv')
tst_newsgroup = pd.read_csv(data_path2, delimiter=',', compression='bz2')
data_path3 = os.path.join(os.getcwd(), 'Data', 'PartA', '20ng_labels.csv')
lbl_newsgroup = pd.read_csv(data_path3, delimiter=',', compression='bz2')
lbl_newsgroup.head
(10)

0
0  comp.sys.ibm.pc.hardware
1     comp.sys.mac.hardware
2                 rec.autos
3           rec.motorcycles
4                 sci.crypt
5           sci.electronics
6    soc.religion.christian
7        talk.religion.misc

In [60]:
# drop class labels
trn_newsgroup_targets = trn_newsgroup['class']
trn_newsgroup = trn_newsgroup.drop(['class'], axis = 1)
tst_newsgroup_targets = tst_newsgroup['class']
tst_newsgroup = tst_newsgroup.drop(['class'], axis = 1)

In [62]:
print('Number of train rows: {}'.format(trn_newsgroup.shape[0]))
print('Number of train columns: {}'.format(trn_newsgroup.shape[1]))
print('Number of test rows: {}'.format(tst_newsgroup.shape[0]))
print('Number of test columns: {}'.format(tst_newsgroup.shape[1]))
trn_newsgroup.describe()
tst_newsgroup.describe()

zeros = 0
ones = 0 
twos = 0
threes = 0
fours = 0
fives = 0
sixes = 0
sevens = 0
for row in trn_newsgroup_targets:
    if(row == 0): 
        zeros = zeros + 1 
    if(row == 1): 
        ones = ones + 1
    if(row == 2): 
        twos = twos + 1
    if(row == 3): 
        threes = threes + 1
    if(row == 4): 
        fours = fours + 1
    if(row == 5): 
        fives = fives + 1
    if(row == 6): 
        sixes = sixes + 1
    if(row == 7): 
        sevens = sevens + 1
print(zeros)
print(ones)
print(twos)
print(threes)
print(fours)
print(fives)
print(sixes)
print(sevens)
print(zeros+ones+twos+threes+fours+fives+sixes+sevens)
trn_newsgroup_targets.head(20)



Number of train rows: 5648
Number of train columns: 1000
Number of test rows: 1883
Number of test columns: 1000
737
722
742
747
743
738
748
471
5648


0     3
1     2
2     1
3     1
4     7
5     4
6     4
7     2
8     6
9     1
10    1
11    0
12    1
13    7
14    3
15    7
16    4
17    3
18    5
19    7
Name: class, dtype: int64

### ==== Question 1.1 =====

Looking at the training set, we see that there are 5648 entries of which 1000 words and their frequencies are being checked. Obviously when dealing with large texts the word frequencies will often be 0 or extremely low. This is the case in our data set as the mean is 0.002292. There are 8 possible classes that the classifier can label an entry as. And from looking at the distribution of these every class seems to be represented quite evenly(~735 each) with the exception of class label 7 which is represented by about 300 less entries. This can lead to the classsifier being more likely to classify a class 7 entry in some other class depending on the algorithm being used for classificiation. Such as one that takes priors into account. 

### ==== Question 1.2 ====

The testing has 1883 entries of newsgroup documents. Resulting in a 75 to 25 percent split in training to testing data. This decision to have significantly more training entries brings its drawbacks and benefits. The larger you make a training set the more information the classifier will have to improve its accuracy during testing and in practice. However, the more testing data you have, the more we understand how accurate the classifier will be in practice. Essentially in prioritising training, we are gaining a more accurate predictor while losing our ability to know how  accurate the classifier will be on previously unseen entries. 

### ==== Question 1.3 ====

Calculating TF-IDF as opposed to just TF is very important, as this accounts for the commonality of a term when determining how important a word is in classifying the document. Common words such as 'is', 'and', 'of', etc are extremely common in all manner of documents, but gives us little to no information on what a documents classification should be.

# 2. Unsupervised Learining

### ==== Question 2.1 =====